From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sing-Han Chen <singhanc@nvidia.com>
Date: Sun, 15 Oct 2023 15:00:51 +0000
Subject: [PATCH 088/245] NVIDIA: SAUCE: usb: xhci: tegra: use reset control
 framework for reset

There is no XUSB powergating support for some internal platforms
Use reset control framework for XUSB controllers.

http://nvbugs/3645739
http://nvbugs/3560606
http://nvbugs/4275301

Change-Id: Iea8e21f0156bce4ea1f30db69662597708c1da7a
Signed-off-by: Sing-Han Chen <singhanc@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2997558
(cherry picked from commit bd07f6991d1bc6ca491d238a437d7a1e7f15eb38)
Signed-off-by: Bodla Rakesh Babu <rbodla@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3178747
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
(cherry picked from commit d4a690c44daef79676b0b99b2a57737d75f24611)

diff --git a/drivers/usb/host/xhci-tegra.c b/drivers/usb/host/xhci-tegra.c
index 98e2b182c0bc4..eb1f2cc75ebf2 100644
--- a/drivers/usb/host/xhci-tegra.c
+++ b/drivers/usb/host/xhci-tegra.c
@@ -250,6 +250,7 @@ struct tegra_xusb_soc {
 	bool otg_set_port_power;
 
 	bool has_bar2;
+	bool has_pg_support;
 };
 
 struct tegra_xusb_context {
@@ -1256,6 +1257,46 @@ static int tegra_xusb_powergate_partitions(struct tegra_xusb *tegra)
 	return 0;
 }
 
+static int tegra_xusb_reset_assert(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+	int rc;
+
+	rc = reset_control_assert(tegra->host_rst);
+	if (rc < 0) {
+		dev_err(dev, "failed to assert xusb_host reset: %d\n", rc);
+		return rc;
+	}
+
+	rc = reset_control_assert(tegra->ss_rst);
+	if (rc < 0) {
+		dev_err(dev, "failed to assert xusb_ss reset: %d\n", rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+static int tegra_xusb_reset_deassert(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+	int rc;
+
+	rc = reset_control_deassert(tegra->host_rst);
+	if (rc < 0) {
+		dev_err(dev, "failed to deassert xusb_host reset: %d\n", rc);
+		return rc;
+	}
+
+	rc = reset_control_deassert(tegra->ss_rst);
+	if (rc < 0) {
+		dev_err(dev, "failed to deassert xusb_ss reset: %d\n", rc);
+		return rc;
+	}
+
+	return 0;
+}
+
 static int __tegra_xusb_enable_firmware_messages(struct tegra_xusb *tegra)
 {
 	struct tegra_xusb_mbox_msg msg;
@@ -1744,7 +1785,7 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 			goto put_padctl;
 		}
 
-		tegra->ss_rst = devm_reset_control_get(&pdev->dev, "xusb_ss");
+		tegra->ss_rst = devm_reset_control_get_shared(&pdev->dev, "xusb_ss");
 		if (IS_ERR(tegra->ss_rst)) {
 			err = PTR_ERR(tegra->ss_rst);
 			dev_err(&pdev->dev, "failed to get xusb_ss reset: %d\n",
@@ -1870,7 +1911,10 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 		}
 	}
 
-	err = tegra_xusb_unpowergate_partitions(tegra);
+	if (tegra->soc->has_pg_support)
+		err = tegra_xusb_unpowergate_partitions(tegra);
+	else
+		err = tegra_xusb_reset_deassert(tegra);
 	if (err)
 		goto free_firmware;
 
@@ -1966,7 +2010,10 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 remove_usb2:
 	usb_remove_hcd(tegra->hcd);
 powergate:
-	tegra_xusb_powergate_partitions(tegra);
+	if (tegra->soc->has_pg_support)
+		tegra_xusb_powergate_partitions(tegra);
+	else
+		tegra_xusb_reset_assert(tegra);
 free_firmware:
 	dma_free_coherent(&pdev->dev, tegra->fw.size, tegra->fw.virt,
 			  tegra->fw.phys);
@@ -1988,7 +2035,10 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 
 static void tegra_xusb_disable(struct tegra_xusb *tegra)
 {
-	tegra_xusb_powergate_partitions(tegra);
+	if (tegra->soc->has_pg_support)
+		tegra_xusb_powergate_partitions(tegra);
+	else
+		tegra_xusb_reset_assert(tegra);
 	tegra_xusb_powerdomain_remove(tegra->dev, tegra);
 	tegra_xusb_phy_disable(tegra);
 	tegra_xusb_clk_disable(tegra);
@@ -2289,7 +2339,12 @@ static int tegra_xusb_enter_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 	if (wakeup)
 		tegra_xhci_enable_phy_sleepwalk_wake(tegra);
 
-	tegra_xusb_powergate_partitions(tegra);
+	if (tegra->soc->has_pg_support)
+		tegra_xusb_powergate_partitions(tegra);
+	else {
+		if (!is_auto_resume)
+			tegra_xusb_reset_assert(tegra);
+	}
 
 	for (i = 0; i < tegra->num_phys; i++) {
 		if (!tegra->phys[i])
@@ -2335,7 +2390,14 @@ static int tegra_xusb_exit_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 		goto out;
 	}
 
-	err = tegra_xusb_unpowergate_partitions(tegra);
+	if (tegra->soc->has_pg_support)
+		err = tegra_xusb_unpowergate_partitions(tegra);
+	else {
+		if (is_auto_resume)
+			err = reset_control_reset(tegra->host_rst);
+		else
+			err = tegra_xusb_reset_deassert(tegra);
+	}
 	if (err)
 		goto disable_clks;
 
@@ -2393,7 +2455,13 @@ static int tegra_xusb_exit_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 		if (!wakeup)
 			phy_exit(tegra->phys[i]);
 	}
-	tegra_xusb_powergate_partitions(tegra);
+
+	if (tegra->soc->has_pg_support)
+		tegra_xusb_powergate_partitions(tegra);
+	else {
+		if (!is_auto_resume)
+			tegra_xusb_reset_assert(tegra);
+	}
 disable_clks:
 	tegra_xusb_clk_disable(tegra);
 out:
@@ -2599,6 +2667,7 @@ static const struct tegra_xusb_soc tegra124_soc = {
 		.owner = 0xf0,
 		.smi_intr = XUSB_CFG_ARU_SMI_INTR,
 	},
+	.has_pg_support = true,
 };
 MODULE_FIRMWARE("nvidia/tegra124/xusb.bin");
 
@@ -2638,6 +2707,7 @@ static const struct tegra_xusb_soc tegra210_soc = {
 		.owner = 0xf0,
 		.smi_intr = XUSB_CFG_ARU_SMI_INTR,
 	},
+	.has_pg_support = true,
 };
 MODULE_FIRMWARE("nvidia/tegra210/xusb.bin");
 
@@ -2683,6 +2753,7 @@ static const struct tegra_xusb_soc tegra186_soc = {
 		.smi_intr = XUSB_CFG_ARU_SMI_INTR,
 	},
 	.lpm_support = true,
+	.has_pg_support = true,
 };
 
 static const char * const tegra194_supply_names[] = {
@@ -2717,6 +2788,7 @@ static const struct tegra_xusb_soc tegra194_soc = {
 		.smi_intr = XUSB_CFG_ARU_SMI_INTR,
 	},
 	.lpm_support = true,
+	.has_pg_support = true,
 };
 MODULE_FIRMWARE("nvidia/tegra194/xusb.bin");
 
@@ -2752,6 +2824,7 @@ static const struct tegra_xusb_soc tegra234_soc = {
 	},
 	.lpm_support = true,
 	.has_bar2 = true,
+	.has_pg_support = true,
 };
 
 static const struct of_device_id tegra_xusb_of_match[] = {
-- 
2.43.0

